### PR TITLE
add more template functions according to fermyon/bartholomew#26 fermy…

### PR DESCRIPTION
**Pull request by karthik2804**
Added more helpers as per fermyon/bartholomew#26, fermyon/bartholomew#27, fermyon/bartholomew#28 & fermyon/bartholomew#31

Signed-off-by: karthik Ganeshram <karthik.ganeshram@fermyon.com>

### CODE REVIEW

File: src/list.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+use handlebars::{handlebars_helper, Handlebars};
+
+pub fn addhelpers(x: &mut Handlebars) {
+    handlebars_helper!(until_step_count: |start: isize, count: isize, step: usize| {
```
Comment (rajatjindal) [inline src/list.rs]: probably rename `count` to `stop` or `max` here. count seems misleading. what do you think? 

also name of the function could be `until_step` instead of `until_step_count`.

please let me know if i misunderstood something here

File: src/template.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+use handlebars::{handlebars_helper, Handlebars};
+use rand::seq::SliceRandom;
+
+pub fn addhelpers(x: &mut Handlebars) {
+    handlebars_helper!(wrap: |index: usize, s: String| {
```
Comment (rajatjindal) [inline src/template.rs]: this will wrap the input just once, so first line will be wrapped as expected, but 2nd line can potentially be longer than wrap index.

I think we should wrap the complete text such that all the new lines are wrapped at given index.

Comment (rajatjindal) [reply src/template.rs]: this also does not handle long words and url's at the moment. golang's version of sprig handles that. you think we can include that functionality as well here?

File: src/template.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        let mut data = s.clone();
+        if s.len() > index + 1 {
+            data.insert(index, '\n')
+        } else {
+            data.push('\n');
```
Comment (rajatjindal) [inline src/template.rs]: I would think it to just return string as it is, if length of string is less than index, but i want to hear your thoughts about it.

File: src/template.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        data
+    });
+    handlebars_helper!(wrap_with: |index: usize, ending: String, s: String| {
+        let mut data = s.clone();
+        if s.len() > index + 1 {
```
Comment (rajatjindal) [inline src/template.rs]: similar to the comment i made in `wrap` function, i think this should wrap all lines as well

File: src/template.rs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                data.insert(index, ch);
+            }
+        } else {
+            for ch in ending.chars() {
+                data.push(ch);
+            }
+        };
+        data
+    });
+    handlebars_helper!(contains: |sub_str: String, content: String | content.contains(&sub_str));
```
Comment (rajatjindal) [inline src/template.rs:5]: hmm, should the following be moved to strings. they sound more like string template functions

**Review by rajatjindal — CHANGES_REQUESTED** [review]
thanks a lot for the PR, the changes look great. I just have a few minor comments.

File: src/list.rs (hunk 1 above)
Comment (karthik2804) [reply src/list.rs]: It makes sense, I will make the changes.

File: src/template.rs (hunk 5 above)
Comment (karthik2804) [reply src/template.rs:5]: I feel like this function will be used in templates in `if` conditions and it does not return a string but a boolean value.

File: src/template.rs (hunk 2 above)
Comment (karthik2804) [reply src/template.rs]: added long word support but not url.

File: src/template.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                        lastwrap = k + ending.len();
+                        i += ending.len() - 1;
+                        break;
+                    }
+                }
+                wraploc = i - lastwrap;
+            }
+            i += 1;
+        }
+        data
```
Comment (rajatjindal) [inline src/template.rs]: is it possible to pull this common code between `wrap` and `wrap_with` into a function and reuse? if we think about it `wrap` is essentially `wrap_with \n`

File: src/template.rs (hunk 2 above)
Comment (rajatjindal) [reply src/template.rs]: thank you. this is fine.

File: src/template.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
+use handlebars::{handlebars_helper, Handlebars};
+use rand::seq::SliceRandom;
+
+pub fn addhelpers(x: &mut Handlebars) {
+    handlebars_helper!(wrap: |index: usize, s: String| {
+        wrap_helper(index, "/n".to_string(), s)
```
Comment (rajatjindal) [inline src/template.rs]: should this be '\n' instead of '/n'?

File: src/template.rs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    x.register_helper("contains", Box::new(contains));
+    x.register_helper("has_prefix", Box::new(has_prefix));
+    x.register_helper("has_suffix", Box::new(has_suffix));
+    x.register_helper("cat", Box::new(cat));
+    x.register_helper("indent", Box::new(indent));
+    x.register_helper("replace", Box::new(replace));
+    x.register_helper("shuffle", Box::new(shuffle));
+}
+
+fn wrap_helper(index: usize, ending: String, s: String) -> String {
```
Comment (rajatjindal) [inline src/template.rs]: could you please write some unit tests for this function. i tried running this through template and got following error:

```
thread 'main' panicked at 'attempt to subtract with overflow', src/template.rs:52:20
```

examples/template.rs

```rust
use handlebars::Handlebars;
use handlebars_sprig;

fn main() {
    // Get a handlebars instance
    let mut hbs = Handlebars::new();

    // THE IMPORTANT PART: Add the helpers
    handlebars_sprig::template::addhelpers(&mut hbs);

    // From this point, you can do whatever you want to do with the
    // handlebars instance. It will have all of the functions available
    // to the templates.
    let tpl = "{{ wrap 5 \"this is a string that needs wrapping\" }}";

    // Example of running a template render.
    let empty_context: Vec<String> = vec![];
    println!(
        "Template produced: {}",
        hbs.render_template(tpl, &empty_context).unwrap()
    )
}
```

Comment (rajatjindal) [reply src/template.rs]: also, please test using unicode characters and verify how the function behaves.

**Review by rajatjindal — CHANGES_REQUESTED** [review]

File: src/template.rs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    let mut data = s.clone();
+    let mut wraploc: usize = 0;
+    let mut lastwrap: usize = 0;
+    let mut i = 0;
+
+    while i < s.len() - 1 {
+        wraploc += 1;
+        if wraploc >= index {
```
Comment (rajatjindal) [inline src/template.rs]: i find the code more readable when there are not too many nested if conditions. you can probably write this as:

```
if wraploc > index {
    i += 1
    continue
}

for k in (0..i).rev() 
.
.
.

**Comment by karthik2804** [comment]
@rajatjindal text wrapping seems to require a bit more logic and testing than what has been presented so far. What would your thoughts be on the deffering adding `wrap` to a later PR (potentially just pulling in a library)?

**Comment by rajatjindal** [comment]
sounds good to me. we can add that in a later PR

**Comment by karthik2804** [comment]
This is ready for a rereview

**Comment by rajatjindal** [comment]
thanks a lot for your PR.

**Comment by karthik2804** [comment]
Would you like to PR the changes into [Bartholomew](https://github.com/fermyon/bartholomew) or should I go ahead and make the PR?

**Comment by rajatjindal** [comment]
if you could pls do it, that would be great.

**Comment by karthik2804** [comment]
Will do it today!
